src/classes/OPP_CreateTribute_BATCH.cls

```diff
@@ -0,0 +1,132 @@
+    private String additionalWhereClause = null;
+
+    /*******************************************************************************************************
+    * @description no argument constructor for the class
+    */
+    public OPP_CreateTribute_BATCH() {
+
+    }
+
+    /*******************************************************************************************************
+    * @description constructor for the class
+    * @param additionalWhereClause - String allowing batch job to be invoked with alternative/additional constraints
+    */ 
+    public OPP_CreateTribute_BATCH(String additionalWhereClause) {
+        this.additionalWhereClause = additionalWhereClause;
+    }
+
+    /******************************************************************************************************
+    * @description Batch start method. Runs through all the Opportunities to be converted.
+    */
+    public Database.QueryLocator start(Database.BatchableContext info) {
+        String soqlQuery = (String.isNotBlank(additionalWhereClause)) ? OPP_QUERY + ' WHERE (' + additionalWhereClause + ')' : OPP_QUERY;
+
+        // Deliberately not trapping this potentially invalid query (since Where Clause might be invalid).
+        // This will display an error in the job widget used on the settings page, and send the person who 
+        // launched it a script error.  Decided this was best option with Nicolas Campbell
+        return Database.getQueryLocator(soqlQuery);
+    }
+
+    /*******************************************************************************************************
+    * @description Batch execute method.
+    * @return void.
+    ********************************************************************************************************/
+    public void execute(Database.BatchableContext BC, List<Opportunity> scope) {
+
+        List<Opportunity> updOpportunities = new List<Opportunity>();
+
+        List<Tribute__c> newTributes = new List<Tribute__c>();
+        for (Opportunity opp : scope) {
+            if (hasTributeInfo(opp)) {
+                newTributes.add(
+                    new Tribute__c(
+                        Opportunity__c = opp.Id,
+                        Tribute_Type__c = opp.Tribute_Type__c,
+                        Honoree_Contact__c = opp.Honoree_Contact__c,
+                        Honoree_Name__c = opp.Honoree_Name__c
+                    )
+                );
+
+                //clear the Opportunity fields
+                opp.Tribute_Type__c = null;
+                opp.Honoree_Contact__c = null;
+                opp.Honoree_Name__c = null;
+                
+                updOpportunities.add(opp);
+            }
+        }
+        
+        if (updOpportunities.size() > 0) {
+            //We do not want to run through Opportunity or Tribute triggers when clearing Tribute Fields
+            //on the Opportunity
+            TDTM_TriggerHandler.disableTDTM = true;
+            insert newTributes;
```

Comment:
Sort of. If there is an error in the batch, the entire batch fails (200 records). The error displayed in the progress component is the error that ApexJobs shows, which is generic saying something along the lines of Record 0 failed. There's no way at all for a User to determine which record failed out of that group of 200; and they'll never really be able to determine which 200 records were not processed. 

There's a pretty strong possibility that orgs with have validation or other rules set on the Opportunity object that could affect older Opportunities being updated by this batch job.  This is a one-time job though (correct?), so I would be OK with leaving the logic for now. I do think this will come back to bite us eventually and we'll need to modify the logic to trap the error(s) and log which Opportunities failed. Until then though, I'm OK with leaving it as is.

Reply:
@force2b - Maybe we create a WI for this one?

Reply:
Good idea. I don't think it needs to be for the initial release. The A/C's on the WI would likely be something like:

- The NPSP Error Log can be used to identify which specific Opportunities failed to update by Id showing the error message associated with the Id.
- If there are any failed Opportunities in the child batch of 200 records, the entire batch of 200 records will not be updated.
- If there are any failed Opportunities in the batch, the insert of tributes is rolled back
- If there are no failures, the Opportunities and Tributes are updated/inserted as appropriate